numcodecs/compat.py

```diff
@@ -27,80 +25,150 @@
 
-        # ensure shapes are compatible
-        if buf.shape != out.shape:
-            if out.flags.f_contiguous:
-                order = 'F'
-            else:
-                order = 'C'
-            buf = buf.reshape(out.shape, order=order)
+    return out
 
-        # copy via numpy
-        np.copyto(out, buf)
 
-    # handle generic buffer destination
-    else:
+def ensure_text(l, encoding='utf-8'):
+    if isinstance(l, text_type):
+        return l
+    else:  # pragma: py3 no cover
+        return text_type(l, encoding=encoding)
 
-        # obtain memoryview of destination
-        dest = memoryview(out)
 
-        # ensure source is 1D
-        if isinstance(buf, np.ndarray):
-            buf = buf.reshape(-1, order='A')
-            # try to match itemsize
-            dtype = 'u%s' % dest.itemsize
-            buf = buf.view(dtype=dtype)
+def ensure_ndarray_from_memory(o, flatten=True):
+    """Convenience function to obtain a numpy ndarray using memory exposed by object `o`,
+    ensuring that no memory copies are made, and that `o` is not an object array.
 
-        # try to copy via memoryview
-        dest[:] = buf
+    Parameters
+    ----------
+    o : bytes-like
+        Any object exposing a memory buffer. On Python 3 this must be an object exposing
+        the new-style buffer interface. On Python 2 this can also be an object exposing
+        the old-style buffer interface.
+    flatten : bool, optional
+        If True, flatten any multi-dimensional inputs into a one-dimensional memoryview.
 
-    return out
+    Returns
+    -------
+    o : ndarray
 
+    """
 
-def ndarray_from_buffer(buf, dtype):
-    if isinstance(buf, np.ndarray):
-        arr = buf.reshape(-1, order='A').view(dtype)
-    else:
-        arr = np.frombuffer(buf, dtype=dtype)
-    return arr
+    if not isinstance(o, np.ndarray):
 
+        # first try to obtain a memoryview or buffer, needed to ensure that we don't
+        # subsequently copy memory when going via np.array()
 
-def ensure_text(l, encoding='utf-8'):
-    if isinstance(l, text_type):
-        return l
-    else:  # pragma: py3 no cover
-        return text_type(l, encoding=encoding)
+        if PY2:  # pragma: py3 no cover
+            # accept objects exposing either old-style or new-style buffer interface
+            try:
+                o = memoryview(o)
+            except TypeError:
+                o = buffer(o)
+
+        else:  # pragma: py2 no cover
+            o = memoryview(o)
+
+        # N.B., this is not documented, but np.array() will accept an object exposing
+        # a buffer interface, and will take a view of the memory rather than making a
+        # copy, preserving type information where present
+        o = np.array(o, copy=False)
+
+    # check for object arrays
+    if o.dtype == object:
+        raise ValueError('object arrays are not supported')
+
+    if flatten:
+
+        # flatten the array to 1 dimension
+        o = o.reshape(-1, order='A')
```

Comment:
Also thought about this as well. Have mixed feelings about including an option for something that is this simple to do externally. Though can certainly see the value in including it for end users. Would be good to weigh the value of including it vs. not.

Reply:
I've removed the option in latest commit.

Reply:
Hmm...guess I'd be tempted not to do any reshaping in here. Mainly thinking of this for two reasons. First it can introduce a copy potentially. Second users may need to do other things that will cause flattening anyways (e.g. `tobytes`).

Reply:
So I think this depends on the scope of the function, i.e., who needs it and why. 

Forget about PY2 for a minute, let's consider just PY3.

The compressor codecs ``ZLib``, ``BZ2`` and ``LZMA`` require an object that exports the new-style buffer interface (e.g., ``memoryview``) with C-contiguous memory. So for them, it is useful to reshape the input, because it ensures a C-contiguous array, whether the input is C-contiguous, F-contiguous or non-contiguous. IIUC if the input is either C- or F-contiguous then calling ``reshape(-1)`` **will not** result in a copy. If the input is non-contiguous then a copy may occur, but then this is unavoidable - the alternative would be to raise an error that the input needs to be contiguous.

However, other codecs may have less stringent requirements. E.g., they may need a numpy array, but may not care if it is contiguous.

So possibly we should have more than one convenience function for obtaining a numpy array, to account for these differences, with some sort of clear naming scheme, so it is clear which function you should use depending on what requirements you have.